subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.rules;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.Task;
+import org.gradle.api.internal.changedetection.state.TaskExecution;
+import org.gradle.util.ChangeListener;
+import org.gradle.util.DiffUtil;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+@NonNullApi
+public abstract class AbstractPropertyNameTaskStateChanges extends SimpleTaskStateChanges {
+
+    private final TaskExecution previous;
+    private final TaskExecution current;
+    private final String title;
+    private final Task task;
+
+    protected AbstractPropertyNameTaskStateChanges(@Nullable TaskExecution previous, TaskExecution current, String title, Task task) {
+        this.previous = previous;
+        this.current = current;
+        this.title = title;
+        this.task = task;
+    }
+
+    protected abstract Map<String, ?> getProperties(TaskExecution execution);
+
+    @Override
+    protected void addAllChanges(final List<TaskStateChange> changes) {
+        if (previous == null) {
+            return;
+        }
```

Comment:
I wonder why we bother creating any of the `TaskStateChanges` objects if `previous` is `null`...

Reply:
I'll look into not creating all those instances when we have no previous execution.

Reply:
I made sure that we do not create all those instances when we have no task history. We also need to do much less null checks.